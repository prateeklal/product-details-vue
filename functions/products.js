const axios = require("axios");

exports.handler = async (event, context, callback) => {
  console.log(event, '\n', context, '\n')
  let getResponse = await axios.get("https://www.westelm.com/services/catalog/v4/category/shop/new/all-new/index.json");
  console.log(getResponse)
  callback(null, {
    statusCode: 200,
    body: JSON.stringify(getResponse.data)
  })
}