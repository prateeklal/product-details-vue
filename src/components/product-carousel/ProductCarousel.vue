<template>
  <div
    class="modal-container"
    @click.self="closeModal"
  >
    <div :class="{ modal: true, 'modal-open': modalOpenAnimation }">
      <figure>
        <img
          v-lazy="currentSlide.href"
          loading="eager"
          :alt="currentSlide.alt"
          :width="currentSlide.width"
          :height="currentSlide.height"
        >
      </figure>
      <button
        aria-label="Close the modal"
        @click="closeModal"
      >
        Close
      </button>
    </div>
    <div class="chevron">
      <button
        v-if="prevSlide"
        class="prev"
        aria-label="Previous Slide"
        @click="
          slideCount--;
          createSlides();
        "
      >
        <img
          :src="prevSlide.href"
          :alt="prevSlide.alt"
        >
        <v-icon name="chevron-left" />
      </button>
      <button
        v-if="nextSlide"
        class="next"
        aria-label="Next Slide"
        @click="
          slideCount++;
          createSlides();
        "
      >
        <img
          :src="nextSlide.href"
          :alt="nextSlide.alt"
        >
        <v-icon name="chevron-right" />
      </button>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    id: {
      type: String,
      default: ""
    },
    images: Array,
    products: {
      type: Array,
      default: () => []
    }
  },

  data() {
    return {
      imgGroup: this.images,
      slideCount: 0,
      currentSlide: "",
      nextSlide: "",
      prevSlide: "",
      modalOpenAnimation: false
    };
  },

  watch: {
    products(products) {
      this.getProductById(products, this.id);
    }
  },

  created() {
    return this.imgGroup && this.createSlides();
  },

  mounted() {
    setTimeout(() => {
      this.modalOpenAnimation = true;
    }, 100);
  },

  methods: {
    closeModal() {
      this.$router.push("/");
    },

    createSlides() {
      this.nextSlide = this.getNextSlide(this.slideCount, this.imgGroup);
      this.prevSlide = this.getPrevSlide(this.slideCount, this.imgGroup);
      this.currentSlide = this.imgGroup[this.slideCount];
    },

    getNextSlide(slideCount, images) {
      const imgLength = images.length;

      if (imgLength > 0 && slideCount < imgLength) {
        return images[slideCount + 1];
      }

      return "";
    },

    getPrevSlide(slideCount, images) {
      return slideCount > 0 ? images[slideCount - 1] : "";
    },

    getProductById(products, id) {
      let product = products.filter(product => id === product.id);
      this.imgGroup = [product[0].hero, ...product[0].images];
      this.createSlides();
    }
  }
};
</script>

<style lang="scss" scoped>
@import "../../scss/_variables";

div.modal-container::before {
  content: "";
  position: absolute;
  top: 0;
  background: rgba(0, 0, 0, 0.7);
  z-index: 3;
  width: 100%;
  height: 100%;
}

.modal-container {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  display: flex;
  height: 100%;
  justify-content: center;
  z-index: 5;
  flex-direction: column;
  align-items: center;
}

.modal {
  position: relative;
  z-index: 5;
  display: flex;
  background-color: #fff;
  flex-direction: column;
  align-items: center;
  max-width: 450px;
  width: 95%;
  max-height: 450px;
  box-shadow: 0 0 22px rgba(0, 0, 0, 0.5);
  border-radius: 5px;
  overflow: auto;
  transition: transform 0.3s ease-in-out;
  transform: scale(0);

  button {
    position: absolute;
    top: 3px;
    right: 5px;
    background-color: transparent;
    border: none;
    font-size: em(11);
    border-bottom: 1px solid $primary-color;
    cursor: pointer;

    &:hover {
      border: none;
    }
  }
}

.modal-open {
  transform: scale(1);
}

h2 {
  font-size: em(16);
  font-weight: 300;
  display: none;
}

.modal img {
  width: 100%;
  height: auto;
  vertical-align: top;
}

figure {
  width: 100%;
}

.chevron {
  position: absolute;
  top: calc(50% - 20px);
  z-index: 5;
  left: 0;
  width: 100%;

  .icon {
    width: 50px;
    color: #fff;
    stroke-width: 1px;
  }

  & > button {
    cursor: pointer;
    background-color: transparent;
    border: none;
    position: absolute;

    > img {
      position: absolute;
      width: 80px;
      top: -(em(14));

      @media (max-width: 767px) {
        display: none;
      }
    }
  }
}

.chevron .prev {
  left: 0;

  > img {
    right: -80px;
  }
}

.chevron .next {
  right: 0;

  > img {
    left: -80px;
  }
}
</style>
