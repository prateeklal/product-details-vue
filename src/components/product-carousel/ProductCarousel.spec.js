import { shallowMount } from '@vue/test-utils';
import ProductCarousel from './ProductCarousel.vue';
import products from './__mocks__/products.json'

describe('ProductCarousel', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(ProductCarousel);
    wrapper.setProps({
      images: products[0].images,
      id: products[0].id,
      products
    });
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should return false if image length is 0', () => {
    let slide = wrapper.vm.getNextSlide(0, {});
    expect(slide).toBeFalsy();
  });

  it('should return next slide if there are 2 or more images', () => {
    let { images } = products[0];
    let slide = wrapper.vm.getNextSlide(0, images);
    expect(slide.href).toBe(images[1].href);
  });

  it('should not add next slide if there are no more images', () => {
    let { images } = products[0];
    let slide = wrapper.vm.getNextSlide(1, images);
    expect(slide).toBeFalsy();
  });

  it('should return previous slide if slide count > 0', () => {
    let { images } = products[0];
    let slide = wrapper.vm.getPrevSlide(2, images);
    expect(slide.href).toBe(images[1].href);
    slide = wrapper.vm.getPrevSlide(1, images);
    expect(slide.href).toBe(images[0].href);
  });

  it('should return false if slide count === 0', () => {
    let { images } = products[0];
    let slide = wrapper.vm.getPrevSlide(0, images);
    expect(slide).toBeFalsy();
  })
});