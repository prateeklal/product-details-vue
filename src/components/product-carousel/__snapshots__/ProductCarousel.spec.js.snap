// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCarousel should render correctly 1`] = `
<div class="modal-container">
  <div class="modal">
    <figure><img loading="eager" alt="" width="363" height="363"></figure> <button aria-label="Close the modal">
      Close
    </button>
  </div>
  <div class="chevron">
    <!----> <button aria-label="Next Slide" class="next"><img src="https://www.westelm.com/weimgs/rk/images/wcm/products/202003/0005/belgian-linen-ladder-stripe-embroidery-duvet-cover-shams-6-m.jpg" alt="">
      <v-icon name="chevron-right"></v-icon>
    </button></div>
</div>
`;
